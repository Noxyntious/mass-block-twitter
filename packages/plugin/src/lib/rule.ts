import {
  z,
  ZodArray,
  ZodBoolean,
  ZodEnum,
  ZodNullable,
  ZodNumber,
  ZodObject,
  ZodOptional,
  ZodString,
  ZodType,
} from 'zod'
import { languages } from './constants/languages'
import { get } from 'es-toolkit/compat'

const ruleDataSchema = z.object({
  user: z
    .object({
      id: z.string(),
      screen_name: z.string(),
      name: z.string(),
      description: z.string().optional(),
      location: z.string().optional(),
      url: z.string().optional(),
      profile_image_url: z.string().optional(),
      created_at: z.string().optional(),
      is_blue_verified: z.boolean().optional(),
      followers_count: z.number().optional(),
      friends_count: z.number().optional(),
      default_profile: z.boolean().optional(),
      default_profile_image: z.boolean().optional(),
    })
    .optional(),
  tweet: z
    .object({
      id: z.string(),
      text: z.string(),
      created_at: z.string(),
      lang: z
        .enum(languages.map((it) => it.value) as [string, ...string[]])
        .optional(),
      media: z
        .array(
          z.object({
            type: z.union([
              z.literal('photo'),
              z.literal('video'),
              z.literal('animated_gif'),
            ]),
            url: z.string(),
          }),
        )
        .optional(),
      conversation_id_str: z.string(),
      in_reply_to_status_id_str: z.string().optional(),
      quoted_status_id_str: z.string().optional(),
    })
    .optional(),
})
export type RuleData = z.infer<typeof ruleDataSchema>

type StringCondition = {
  field: string
  operator: 'eq' | 'neq' | 'cont' | 'notCont' | 'regex'
  value: string
}
type NumberCondition = {
  field: string
  operator: 'eq' | 'neq' | 'gt' | 'gte' | 'lt' | 'lte'
  value: number
}
type BooleanCondition = {
  field: string
  operator: 'eq'
  value: boolean
}
type Condition = StringCondition | NumberCondition | BooleanCondition
export interface Rule {
  or: {
    and: Condition[]
  }[]
}

type RuleField =
  | {
      field: string
      type: 'string'
      operator: {
        value: StringCondition['operator']
        label: string
      }[]
      enum?: {
        value: string
        label: string
      }[]
    }
  | {
      field: string
      type: 'number'
      operator: {
        value: NumberCondition['operator']
        label: string
      }[]
    }
  | {
      field: string
      type: 'boolean'
      operator: {
        value: BooleanCondition['operator']
        label: string
      }[]
    }

export function visit(
  schema: ZodType,
  iter: (value: ZodType, path: string[]) => void | false,
) {
  function f(schema: unknown, path: string[]) {
    if (!(schema instanceof ZodType)) {
      return
    }
    if (schema instanceof ZodOptional || schema instanceof ZodNullable) {
      f(schema.unwrap(), path)
      return
    }
    if (path.length !== 0) {
      const result = iter(schema, path)
      if (result === false) {
        return
      }
    }
    if (schema instanceof ZodObject) {
      Object.entries(schema.shape).forEach(([key, value]) => {
        f(value, [...path, key])
      })
      return
    }
    if (
      schema instanceof ZodBoolean ||
      schema instanceof ZodNumber ||
      schema instanceof ZodString
    ) {
      return
    }

    throw new Error(`Unsupported schema: ${schema.constructor.name}`)
  }
  f(schema, [])
}

export function getRuleFileds(): RuleField[] {
  const fields: RuleField[] = []
  visit(ruleDataSchema, (it, path) => {
    if (it instanceof ZodArray) {
      return false
    }
    if (it instanceof ZodEnum) {
      if (path.join('.') !== 'tweet.lang') {
        throw new Error('Enum field must be tweet.lang')
      }
      fields.push({
        field: path.join('.'),
        type: 'string',
        operator: [{ value: 'eq', label: 'Equal' }],
        enum: languages,
      })
      return false
    }
    if (it instanceof ZodString) {
      fields.push({
        field: path.join('.'),
        type: 'string',
        operator: [
          { value: 'eq', label: 'Equal' },
          { value: 'neq', label: 'Not Equal' },
          { value: 'cont', label: 'Contains' },
          { value: 'notCont', label: 'Not Contains' },
          { value: 'regex', label: 'Regex' },
        ],
      })
      return
    }
    if (it instanceof ZodNumber) {
      fields.push({
        field: path.join('.'),
        type: 'number',
        operator: [
          { value: 'eq', label: 'Equal' },
          { value: 'neq', label: 'Not Equal' },
          { value: 'gt', label: 'Greater Than' },
          { value: 'gte', label: 'Greater Than or Equal' },
          { value: 'lt', label: 'Less Than' },
          { value: 'lte', label: 'Less Than or Equal' },
        ],
      })
    }
    if (it instanceof ZodBoolean) {
      fields.push({
        field: path.join('.'),
        type: 'boolean',
        operator: [{ value: 'eq', label: 'Equal' }],
      })
    }
  })
  return fields
}

function matchCondition(cond: Condition, data: any) {
  const { field, operator, value } = cond
  const fieldValue = get(data, field)
  if (fieldValue === undefined || fieldValue === null) {
    return false
  }
  switch (operator) {
    case 'eq':
      return fieldValue === value
    case 'neq':
      return fieldValue !== value
    case 'cont':
      return fieldValue.includes(value)
    case 'notCont':
      return !fieldValue.includes(value)
    case 'gt':
      return fieldValue > value
    case 'gte':
      return fieldValue >= value
    case 'lt':
      return fieldValue < value
    case 'lte':
      return fieldValue <= value
    case 'regex':
      if (typeof value === 'string') {
        return new RegExp(value).test(fieldValue)
      }
      return false
    default:
      console.error(`Unknown operator: ${operator}`)
  }
  return false
}

export function matchRule(rules: Rule[], data: any) {
  return rules.some((rule) =>
    rule.or.some((or) => or.and.every((cond) => matchCondition(cond, data))),
  )
}
