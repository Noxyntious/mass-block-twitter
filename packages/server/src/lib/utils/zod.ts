import { z, ZodTypeAny } from 'zod'

/**
 * @link https://github.com/colinhacks/zod/issues/2461#issuecomment-1712769467
 * @param zodPipe
 * @returns
 */
export const zodStringNumber = () =>
  z
    .string()
    .optional()
    .refine((value) => !value || !isNaN(Number(value)), {
      message: 'Invalid number',
    })
    .transform((value) => {
      return value ? Number(value) : undefined
    })
    .pipe(z.number())
