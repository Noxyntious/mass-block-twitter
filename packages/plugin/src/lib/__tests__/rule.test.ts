import { getRuleFileds, visit } from '$lib/rule'
import { describe, expect, it } from 'vitest'
import { z, ZodArray, ZodEnum } from 'zod'

describe('getRuleFileds', () => {
  it('should return the correct fields', () => {
    const fields = getRuleFileds()
    expect(fields).toMatchSnapshot()
  })
})

describe('visit', () => {
  it('should visit the correct fields', () => {
    const fields: string[] = []
    visit(
      z.object({
        a: z.string(),
        b: z.object({
          c: z.number(),
        }),
        d: z.array(z.boolean()),
      }),
      (it, path) => {
        if (it instanceof ZodArray) {
          return false
        }
        fields.push(path.join('.'))
      },
    )
    expect(fields).toEqual(['a', 'b', 'b.c'])
  })
  it('should visit optional fields', () => {
    const fields: string[] = []
    visit(
      z.object({
        a: z.enum(['a', 'b']).optional(),
      }),
      (it, path) => {
        fields.push(path.join('.'))
        if (it instanceof ZodEnum) {
          return false
        }
      },
    )
    expect(fields).toEqual(['a'])
  })
})
