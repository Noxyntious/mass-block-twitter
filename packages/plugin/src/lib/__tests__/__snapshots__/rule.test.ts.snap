// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getRuleFileds > should return the correct fields 1`] = `
[
  {
    "field": "user.id",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "field": "user.screen_name",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "field": "user.name",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "field": "user.description",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "field": "user.location",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "field": "user.profile_image_url",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "field": "user.created_at",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "field": "user.is_blue_verified",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
    ],
    "type": "boolean",
  },
  {
    "field": "user.followers_count",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Greater Than",
        "value": "gt",
      },
      {
        "label": "Greater Than or Equal",
        "value": "gte",
      },
      {
        "label": "Less Than",
        "value": "lt",
      },
      {
        "label": "Less Than or Equal",
        "value": "lte",
      },
    ],
    "type": "number",
  },
  {
    "field": "user.friends_count",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Greater Than",
        "value": "gt",
      },
      {
        "label": "Greater Than or Equal",
        "value": "gte",
      },
      {
        "label": "Less Than",
        "value": "lt",
      },
      {
        "label": "Less Than or Equal",
        "value": "lte",
      },
    ],
    "type": "number",
  },
  {
    "field": "user.default_profile",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
    ],
    "type": "boolean",
  },
  {
    "field": "user.default_profile_image",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
    ],
    "type": "boolean",
  },
  {
    "field": "tweet.id",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "field": "tweet.text",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "field": "tweet.created_at",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "enum": [
      {
        "label": "Arabic",
        "value": "ar",
      },
      {
        "label": "Bengali",
        "value": "bn",
      },
      {
        "label": "Chinese",
        "value": "zh",
      },
      {
        "label": "Czech",
        "value": "cs",
      },
      {
        "label": "Danish",
        "value": "da",
      },
      {
        "label": "Dutch",
        "value": "nl",
      },
      {
        "label": "English",
        "value": "en",
      },
      {
        "label": "Finnish",
        "value": "fi",
      },
      {
        "label": "French",
        "value": "fr",
      },
      {
        "label": "German",
        "value": "de",
      },
      {
        "label": "Greek",
        "value": "el",
      },
      {
        "label": "Hebrew",
        "value": "he",
      },
      {
        "label": "Hindi",
        "value": "hi",
      },
      {
        "label": "Hungarian",
        "value": "hu",
      },
      {
        "label": "Indonesian",
        "value": "id",
      },
      {
        "label": "Italian",
        "value": "it",
      },
      {
        "label": "Japanese",
        "value": "ja",
      },
      {
        "label": "Korean",
        "value": "ko",
      },
      {
        "label": "Malay",
        "value": "ms",
      },
      {
        "label": "Norwegian",
        "value": "no",
      },
      {
        "label": "Persian",
        "value": "fa",
      },
      {
        "label": "Polish",
        "value": "pl",
      },
      {
        "label": "Portuguese",
        "value": "pt",
      },
      {
        "label": "Romanian",
        "value": "ro",
      },
      {
        "label": "Russian",
        "value": "ru",
      },
      {
        "label": "Spanish",
        "value": "es",
      },
      {
        "label": "Swahili",
        "value": "sw",
      },
      {
        "label": "Swedish",
        "value": "sv",
      },
      {
        "label": "Thai",
        "value": "th",
      },
      {
        "label": "Turkish",
        "value": "tr",
      },
      {
        "label": "Ukrainian",
        "value": "uk",
      },
      {
        "label": "Urdu",
        "value": "ur",
      },
      {
        "label": "Vietnamese",
        "value": "vi",
      },
    ],
    "field": "tweet.lang",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
    ],
    "type": "string",
  },
  {
    "field": "tweet.conversation_id_str",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "field": "tweet.in_reply_to_status_id_str",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
  {
    "field": "tweet.quoted_status_id_str",
    "operator": [
      {
        "label": "Equal",
        "value": "eq",
      },
      {
        "label": "Not Equal",
        "value": "neq",
      },
      {
        "label": "Contains",
        "value": "cont",
      },
      {
        "label": "Not Contains",
        "value": "notCont",
      },
    ],
    "type": "string",
  },
]
`;
